README.rst
=========
Trimshare
=========
Trimshare *trims* your videos into short clips, made web-friendly for easy *sharing*. No need to remember the correct FFmpeg incantation!

Trimshare is handy if you have long videos like gameplay footage and want to grab specific moments to share with others in chatrooms or forums. A 10-30 second clip at 720p 60 fps will usually take less than 8 MB.

Details
-------
This command-line tool is a frontend to FFmpeg with a simpler, opinionated interface specifically for extracting short clips from long videos and encoding them into VP9 ``.webm`` files. The VP9 encoder is set to use constant quality mode with a CRF of 50 by default.

Building from source
--------------------
Prerequisites
=============
* pip
* `Poetry <https://python-poetry.org/>`_
* FFmpeg
    * Must be available on your PATH
    * Must be built with VP9 encode support (it very likely already is)

Build
=====
Build and install the ``trimshare`` binary to your PATH::

    poetry install

Usage
-----
Run with the command::

    trimshare

Get full help text for arguments with::

    trimshare --help
    
Example typical usage to grab 0:23 to 0:49 on a file called ``example.mkv``::

    trimshare example.mkv -s 0:23 -e 0:49

This generates a file called ``example.webm`` that's a shortened, compressed clip of the original video file.

License
-------
Copyright © 2022-2023 JonLiuFYI.

Trimshare is free software. Its source is released under the GNU GPL v3. See ``LICENSE``.

CHANGELOG.rst
=========
Changelog
=========

0.1.0 (2023-01-27)
------------------
Initial public release

* Choose output video file name or infer one from original file name
* Choose clip start/end times. If only one is given, assume the other clip boundary time is the start or end of the video.
* Choose CRF quality value, 0-63, lower is better. Default is 50.
* Output a WebM file with a VP9-encoded video stream


tests/test_infer_out_video_name.py
import pytest
from freezegun import freeze_time
from trimshare import __version__
from trimshare.trimshare import infer_out_video_name


def test_version():
    assert __version__ == "0.1.0"


@freeze_time("2001-01-01 01:01:01")
def test_infer_out_video_name(tmp_path):
    # don't infer name: valid, explicit out name
    assert infer_out_video_name("pwn3d.webm", "myvideo.mp4") == "pwn3d.webm"
    # deep input path
    assert (
        infer_out_video_name("pwn3d.webm", "deep/as/heck/path/myvideo.mp4")
        == "pwn3d.webm"
    )
    # same, but infer name
    assert (
        infer_out_video_name(None, "deep/as/heck/path/myvideo.mp4")
        == "deep/as/heck/path/myvideo.webm"
    )

    # infer name, no name conflict
    assert infer_out_video_name(None, "myvideo.mp4") == "myvideo.webm"
    # infer name, lots of dots
    assert infer_out_video_name(None, "my.special.video.mp4") == "my.special.video.webm"
    # infer name, name conflict: webm exists already
    (tmp_path / "myvideo.webm").touch()
    assert infer_out_video_name(None, str(tmp_path / "myvideo.webm")) == str(
        tmp_path / "myvideo-trimshare-2001-01-01-010101.webm"
    )


def test_same_out_and_in_name_error():
    with pytest.raises(FileExistsError):
        infer_out_video_name("myvideo.webm", "myvideo.webm")


@freeze_time("2001-01-01 01:01:01")
def test_rare_name_conflict(tmp_path):
    (tmp_path / "stuff.mkv").touch()
    (tmp_path / "stuff.webm").touch()
    (tmp_path / "stuff-trimshare-2001-01-01-010101.webm").touch()
    with pytest.raises(NameError):
        infer_out_video_name(None, str(tmp_path / "stuff.mkv"))


trimshare/trimshare.py
#!/usr/bin/env python3
"""trimshare.py: Trim a video and convert to webm for easy sharing.

trimshare makes short clips out of long video files. It then converts
them to webm with the vp9 codec. By default, the quality is CRF 50.
At 720p and 60 fps, a clip of 10-30 seconds is usually less than 8 MB.

Source code of trimshare is Copyright © 2022-2023 JonLiuFYI,
licensed under GNU GPL v3. See LICENSE.

Example to grab 0:23-0:49 of `example.mkv`:
    trimshare example.mkv -s 0:23 -e 0:49

Usage:
    trimshare.py <in_video> [options]
    trimshare.py -h | --help
    trimshare.py --version

Arguments:
    <in_video>              Path to the video to trim.

Options:
    -o <out_video>          Export trimmed clip to this path as webm. If unspecified, automatically choose a name similar to <in_video> and save the clip in the same directory.
    -s --starttime <t>      Start the clip at this point in time. If unspecified, start at 0:00 on the source video.
    -e --endtime <t>        End the clip at this point in time. If unspecified, stop at the end of the source video.
    -q --quality <qual>     Set output vp9 quality (CRF mode), 0-63. Lower is better. [default: 50]
    --debug                 Show debugging info while this runs.
    -h --help               Show this help info.
    --version               Show the version number of trimshare.
"""
import logging
from pathlib import Path
import subprocess as sp

from docopt import docopt

VERSION = "trimshare 0.1.0"


def infer_out_video_name(outname: str | None, inname: str) -> str:
    """Determine what the output file's name should be, if needed.

    If ``outname`` is None, return the input video's file name with the
    extension changed to ``.webm``, adding ``-trimshare-{datetime}`` to
    avoid conflicts if needed.

    Args:
        outname (str | None): Output file name or None. If None, make up a name.
        inname (str): Input file name

    Raises:
        FileExistsError: ``outname`` and ``inname`` are the same

    Returns:
        str: File name to output to
    """
    if outname is not None:
        if outname == inname:
            raise FileExistsError("Output video must be different from the input video")
        return outname

    # strip file extension if there is one
    if "." in inname:
        inname = ".".join(inname.split(".")[:-1])

    conv_name = f"{inname}.webm"
    if Path(conv_name).exists():
        from datetime import datetime

        now = datetime.now().strftime("%Y-%m-%d-%H%M%S")
        safer_name = f"{inname}-trimshare-{now}.webm"
        if Path(safer_name).exists():
            raise NameError(
                f"Could not automatically generate a safe file name for {inname}"
            )
        return safer_name
    else:
        return conv_name


def main():
    args = docopt(__doc__, version=VERSION, more_magic=False)  # type: ignore
    logging.basicConfig(level=logging.DEBUG if args.debug else logging.ERROR)

    in_video = args.in_video
    if not Path(in_video).exists():
        raise SystemExit(f"No file called {in_video}")

    start_time = args.starttime
    end_time = args.endtime
    out_video = infer_out_video_name(args.o, in_video)
    qual = args.quality
    logging.debug(
        "Given args:\n"
        f"\t{in_video = }\n"
        f"\t{start_time = }\n"
        f"\t{end_time = }\n"
        f"\t{out_video = }\n"
        f"\t{qual = }"
    )

    start_time_arg = ["-ss", start_time] if start_time is not None else []
    end_time_arg = ["-to", end_time] if end_time is not None else []

    # fmt: off
    pass1 = [
        "ffmpeg",
        "-i", in_video,
        *start_time_arg,
        *end_time_arg,
        "-c:v", "libvpx-vp9",
        "-b:v", "0",
        "-crf", qual,
        "-row-mt", "1",
        "-pass", "1",
        "-an",
        "-f", "null",
        "/dev/null",
    ]
    pass2 = [
        "ffmpeg",
        "-i", in_video,
        *start_time_arg,
        *end_time_arg,
        "-c:v", "libvpx-vp9",
        "-b:v", "0",
        "-crf", qual,
        "-row-mt", "1",
        "-pass", "2",
        out_video,
    ]
    # fmt: on
    logging.debug(f"Pass 1: {' '.join(pass1)}")
    logging.debug(f"Pass 2: {' '.join(pass2)}")

    firstpass = sp.run(pass1)
    if firstpass.returncode != 0:
        raise SystemExit(
            "The first encoding pass failed! Check ffmpeg's output for the reason why."
        )
    sp.run(pass2)


if __name__ == "__main__":
    main()
